staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic.go

```diff
@@ -0,0 +1,292 @@
+	delegate      audit.Backend
+	stopChan      chan struct{}
+}
+
+// NewBackend returns a backend that dynamically updates its configuration
+// based on a shared informer.
+func NewBackend(c *Config) (audit.Backend, error) {
+	informer := c.F.Auditregistration().V1alpha1().AuditSinks()
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventSink := &v1core.EventSinkImpl{Interface: c.KubeClient.CoreV1().Events("")}
+	eventBroadcaster.StartRecordingToSink(eventSink)
+
+	scheme := runtime.NewScheme()
+	err := auditregv1alpha1.AddToScheme(scheme)
+	if err != nil {
+		return nil, err
+	}
+	recorder := eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: "apiserver"})
+
+	if c.BufferedConfig == nil {
+		defaultConfig := NewDefaultWebhookBatchConfig()
+		c.BufferedConfig = &defaultConfig
+	}
+
+	cm, err := webhook.NewClientManager(auditinternal.SchemeGroupVersion, auditinternal.AddToScheme)
+	if err != nil {
+		return nil, err
+	}
+	authInfoResolver, err := webhook.NewDefaultAuthenticationInfoResolver("")
+	if err != nil {
+		return nil, err
+	}
+	cm.SetAuthenticationInfoResolver(authInfoResolver)
+	cm.SetServiceResolver(webhook.NewDefaultServiceResolver())
+	cm.SetAuthenticationInfoResolverWrapper(c.AuthInfoResolverWrapper)
+
+	manager := &dynamic{
+		config:               c,
+		backends:             &atomic.Value{},
+		backendUpdateMutex:   sync.Mutex{},
+		webhookClientManager: cm,
+		recorder:             recorder,
+	}
+	manager.backends.Store(syncedBackends{})
+
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			manager.addSink(obj.(*auditregv1alpha1.AuditSink))
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			manager.updateSink(oldObj.(*auditregv1alpha1.AuditSink), newObj.(*auditregv1alpha1.AuditSink))
+		},
+		DeleteFunc: func(obj interface{}) {
+			manager.deleteSink(obj.(*auditregv1alpha1.AuditSink))
+		},
+	})
+
+	return manager, nil
+}
+
+type dynamic struct {
+	// backendUpdateMutex holds an update lock on the backends
+	backendUpdateMutex   sync.Mutex
+	config               *Config
+	backends             *atomic.Value
+	webhookClientManager webhook.ClientManager
+	recorder             record.EventRecorder
+}
+
+type syncedBackends map[types.UID]*Backend
+
+// Names returns the names of the backend configurations
+func (s syncedBackends) Names() []string {
+	names := []string{}
+	for _, backend := range s {
+		names = append(names, backend.configuration.Name)
+	}
+	return names
+}
+
+// ProcessEvents proccesses the given events per current backend map
+func (d *dynamic) ProcessEvents(events ...*auditinternal.Event) {
+	for _, b := range d.GetBackends() {
+		b.delegate.ProcessEvents(events...)
+	}
+}
+
+// Run the dynamic backends
+func (d *dynamic) Run(stopCh <-chan struct{}) error {
+	// Shutdown channels are handled individually with the dynamic
+	// backend, message must be propagated out
+	go func() {
```

Comment:
does doing this first make it possible for stopAllBackends to be called before b.delegate.Run is called?